src/Juvix/Compiler/Core/Transformation/LambdaLifting.hs

```diff
@@ -24,48 +21,45 @@ lambdaLiftNode aboveBl top =
     typeFromArgs :: [ArgumentInfo] -> Type
     typeFromArgs = \case
       [] -> mkDynamic' -- change this when we have type info about the body
-      (a : as) -> mkPi' argTy (typeFromArgs as)
-        where
-          argTy = fromMaybe mkDynamic' (a ^. argumentType)
+      (a : as) -> mkPi' (a ^. argumentType) (typeFromArgs as)
```

Comment:
Do you remember that Pi introduces a binder? It seems this may be incorrect with (type) variables occurring in types. The target of Pi may need to be shifted to adjust its context. This depends on how we create ArgumentInfo, i.e., what are the contexts of types in it. The initial idea was that it doesn't matter, because initially there were no explicit binders in types. Now we need to decide something, e.g., the context of the first argument is global (closed), the context of the second contains the first variable, etc., then the types in ArgumentInfo correspond to the argument types obtained from unfoldType. I think this is the current decision in JVC parser, but there the function is already top level so its type must be closed. With this decision you don't need shifting here unless the type you create ArgumentInfo from was not closed in the first place.

The proper way to do this is to close also under the free variables occurring in argument types, and recursively so. We may defer this, but it's an issue -- otherwise it won't work with polymorphic functions as polymorphism is explicit now.

Reply:
Afaik currently recursors don't look at the types of the nodes, so `getFreeVars` does not return free type variables in the type of some node. This could be fixed by changing the behavriour of the recursor, right?
Once we have a function that returns all free variables (including those appearing in types), it should be enough to close the terms with the sorted variables.I.e. the only required change would be
`captureFreeVars freevarsAssocs l'`
`captureFreeVars (sortByIx freevarsAssocs) l'`
Does this make sense?

Reply:
p.s. I think we should make this change in a separate pr, so I propose that we merge this and then tackle this issue

Reply:
I think that would work for polymorphism, but not for dependent types. Consider:
```
\(A : Type) \(B : T A) (\x : B) f (\(y : B) h y x)
```
The result of lambda-lifting should be:
```
def lam := \(A : Type) \(B : T A) \(x : B) \(y : B) h y x

\(A : Type) \(B : T A) (\x : B) f (lam A B x)
```
But `A` does not occur anywhere in the expression being lifted. To make it work in general one needs to close recursively until there are no more free variables (but I think one could do a single closure the way you propose).

Yes, we can change it in a separate PR.